feat: Amazon Linux 2023 support, provider updates, and security enhancements

prabirsekhri opened the pull request:
## PR Title
**feat: Amazon Linux 2023 support, provider updates, and security enhancements (v3.0.0)**

---

## Description

This PR brings significant updates to support Amazon Linux 2023-based Storage Gateway deployments and modernizes the module with updated provider requirements and security enhancements.

### Breaking Changes

| Component | Previous | New |
|-----------|----------|-----|
| Terraform | >= 1.3.0 | >= 1.5.7 |
| AWS Provider | >= 4.0.0 | >= 5.0.0 |
| AWSCC Provider | >= 0.25.0 | >= 1.0.0 |
| vSphere Provider | - | >= 2.2.0 |
| S3 Bucket Module | >=3.5.0 | ~> 5.0 |

### New Features

#### Amazon Linux 2023 Support
- **EC2 module**: Updated AMI lookup to use SSM parameters for AL2023-based Storage Gateway AMIs with support for all gateway types (FILE_S3, VTL, CACHED, STORED)
- **VMware module**: Updated default OVA URL to v2 (AL2023): `https://dd958of58tzpr.cloudfront.net/aws-storage-gateway-file-s3-gateway-v2-x86_64.ova`
- Added `vsphere_ovf_vm_template` data source to properly detect guest OS properties (fixes "Other Linux" detection issue on vSphere 7+)

#### Security Enhancements
- Enhanced EC2 `metadata_options` with explicit IMDSv2 settings:
  - `http_endpoint = "enabled"`
  - `http_tokens = "required"`
  - `http_put_response_hop_limit = 1`
- Refactored security groups to use separate `aws_security_group_rule` resources
- Added missing egress rule for VPC endpoint security groups

#### Bug Fixes
- Made `domain_controllers` optional for AD join (fixes #16)
- Added lifecycle rule to ignore AMI changes and prevent unexpected instance replacement

### Code Quality
- Updated `.tflint.hcl`: Changed deprecated `module = true` to `call_module_type = "all"`
- Updated AWS tflint plugin from `0.18.0` to `0.32.0`
- Removed unused `timezone` variable from ec2-sgw module
- Fixed typo in vmware-sgw documentation
- Synced root module with aws-sgw submodule for consistency

### Related Issues/PRs
- Supersedes #15 (metadata params fix)
- Supersedes #16 (domain controllers length fix)
- Addresses concerns from #18 (optional public IP - partial)

### Migration Guide

Users upgrading from v2.x should:
1. Update Terraform to version 1.5.7 or later
2. Update AWS provider to version 5.0.0 or later
3. Run `terraform init -upgrade` to update module dependencies
4. Review any custom security group configurations

### Testing
- [x] tflint passes (`tflint --recursive`)
- [x] VMware deployment tested with vSphere 7
- [x] EC2 deployment tested with AL2023 AMI

wellsiau-aws commented on README.md:
would be nice to have some Breaking Change section in this readme to warn user for migration path.

prabirsekhri replied:
Added

wellsiau-aws commented on examples/s3-smb-filegateway-vmware/variables.tf:
```diff
@@ -29,6 +29,13 @@ variable "domain_controllers" {
 
+variable "organizational_unit" {
+  type        = string
+  default     = ""
```
is it intentional to set the value to "" rather than leaving it as null? My naive question is without looking at the actual code where this var is used.

prabirsekhri replied:
good catch, updated to null values instead of empty strings for all variables where this was applicable

wellsiau-aws commented on modules/ec2-sgw/main.tf:
```diff
@@ -35,21 +49,15 @@ resource "aws_instance" "ec2_sgw" {
-    values = ["aws-storage-gateway-FILE_S3*"]
+    # Ignore AMI changes to prevent unexpected instance replacement
+    # To update the AMI, use: terraform apply -replace="module.ec2_sgw.aws_instance.ec2_sgw"
+    ignore_changes = [ami]
```
any reason why this would be the default behavior?

prabirsekhri replied:
This is done to address this PR https://github.com/aws-ia/terraform-aws-storagegateway/pull/20. Ideally we want customers to plan their storage gateway upgrades as it can be a disruptive operation.

wellsiau-aws reviewed (commented):
I added a few questions, overall looks good to me.

SamuelOuvrard commented on .config/.tfsec/launch_configuration_imdsv2_tfchecks.json:
Is this still necessary? I thought tfsec was being deprecated. https://github.com/aquasecurity/tfsec

prabirsekhri replied:
good point, removed

SamuelOuvrard commented on .config/.tfsec/launch_template_imdsv2_tfchecks.json:
Same comment about tfsec being deprecated

prabirsekhri replied:
removed

SamuelOuvrard commented on .github/workflows/static-tests.yml:
If you want this to run similarly to the other IA projects managed by APB this action should launch on push to any branch, not just a push or PR to main.

prabirsekhri replied:
done

harsheetas commented on modules/ec2-sgw/variables.tf:
```diff
@@ -35,32 +54,39 @@ variable "ingress_cidr_blocks" {
-  type        = string
-  description = "Time zone for the gateway. The time zone is of the format GMT, GMT-hr:mm, or GMT+hr:mm.For example, GMT-4:00 indicates the time is 4 hours behind GMT. Avoid prefixing with 0"
-  default     = "GMT"
+  description = "The instance type to use for the Storage Gateway. Instance types supported are m5.xlarge (minimum for small deployment), m5.2xlarge (medium), or m5.4xlarge (large). M6i variants are also supported."
```
For supported instance types we can add a [ref](https://docs.aws.amazon.com/filegateway/latest/files3/Requirements.html#requirements-hardware-ec2) for all supported instance families instead of restricting it to just m5 and m6.

harsheetas commented on modules/ec2-sgw/variables.tf:
```diff
@@ -35,32 +54,39 @@ variable "ingress_cidr_blocks" {
   validation {
-    condition     = can(regex("^GMT[+-](([1-9]|1[0-2]):([0-5][0-9]))|GMT$", var.timezone))
-    error_message = "Time zone for the gateway. The time zone is of the format GMT, GMT-hr:mm, or GMT+hr:mm."
+    condition     = contains(["m5.xlarge", "m5.2xlarge", "m5.4xlarge", "m6i.xlarge", "m6i.2xlarge", "m6i.4xlarge"], var.instance_type)
```
We can leave the default instance type to m5.xlarge but remove the validation as it limits the instance types supported.

SamuelOuvrard commented on modules/aws-sgw/sg.tf:
```diff
@@ -1,4 +1,5 @@
+#checkov:skip=CKV2_AWS_5:Security group is conditionally created and attached to VPC endpoint when create_vpc_endpoint=true
 resource "aws_security_group" "vpce_sg" {
 
   for_each = (var.create_vpc_endpoint && var.create_vpc_endpoint_security_group) ? toset(["vpce_sg"]) : toset([])
```
Is there ever a version of this where we create more than a single security group? If not should we be using count instead of for_each?

SamuelOuvrard commented on modules/aws-sgw/sg.tf:
```diff
@@ -58,3 +59,17 @@ resource "aws_security_group_rule" "vpce_2222" {
+#tfsec:ignore:aws-ec2-no-public-egress-sgr
+resource "aws_security_group_rule" "vpce_egress" {
+  #checkov:skip=CKV_AWS_382:Storage Gateway VPC endpoint requires outbound access to AWS services
+  for_each          = (var.create_vpc_endpoint && var.create_vpc_endpoint_security_group) ? toset(["vpce_egress"]) : toset([])
```
same here?

prabirsekhri replied:
resolved

SamuelOuvrard commented on modules/aws-sgw/variables.tf:
should there be any validation for the variables domain_name, domain_username, domain_password or organizational_unit?

prabirsekhri replied:
These variables are only used conditionally when join_smb_domain=true. The AWS API provides clear validation errors, and domain/OU formats vary enough that strict validation could reject valid inputs.

SamuelOuvrard commented on modules/ec2-sgw/main.tf:
```diff
@@ -4,10 +4,26 @@
+  name = local.gateway_type_ssm_paths[var.gateway_type]
 }
 
+#checkov:skip=CKV_AWS_126:Detailed monitoring is optional for Storage Gateway EC2 instances
```
we shouldn't have the skip comments twice

prabirsekhri replied:
good catch, resolved

SamuelOuvrard commented on modules/ec2-sgw/sg.tf:
```diff
@@ -4,6 +4,7 @@ locals {
+#checkov:skip=CKV2_AWS_5:Security group is conditionally created and attached to EC2 instance when create_security_group=true
 resource "aws_security_group" "ec2_sg" {
 
   for_each = var.create_security_group == true ? toset(["ec2_sg"]) : toset([])
```
same comment about for_each vs count

SamuelOuvrard replied:
if this changes to count instead of for_each, please test without the checkov skip

prabirsekhri replied:
resolved

SamuelOuvrard commented on modules/ec2-sgw/sg.tf:
```diff
@@ -13,6 +14,8 @@ resource "aws_security_group" "ec2_sg" {
 }
 
+#HTTP required for activation - CIDR is configurable via ingress_cidr_block_activation variable
+#tfsec:ignore:aws-ec2-no-public-ingress-sgr
```
should we still be running tfsec at all?

prabirsekhri replied:
agreed, removed

SamuelOuvrard replied:
Then the `tfsec:ignore` can be removed as well!

harsheetas reviewed (commented):
Added feedback regarding the instance family choice for EC2 gateways. 
Otherwise, looks good to merge.

SamuelOuvrard commented on sg.tf:
```diff
@@ -1,4 +1,5 @@
+#checkov:skip=CKV2_AWS_5:Security group is conditionally created and attached to VPC endpoint when create_vpc_endpoint=true
 resource "aws_security_group" "vpce_sg" {
 
   for_each = (var.create_vpc_endpoint && var.create_vpc_endpoint_security_group) ? toset(["vpce_sg"]) : toset([])
```
same as before for_each vs count

prabirsekhri replied:
resolved

SamuelOuvrard commented on sg.tf:
```diff
@@ -13,58 +14,62 @@ resource "aws_security_group" "vpce_sg" {
 
-  ingress {
+resource "aws_security_group_rule" "vpce_443" {
+  for_each          = (var.create_vpc_endpoint && var.create_vpc_endpoint_security_group) ? toset(["vpce_443"]) : toset([])
```
same

prabirsekhri committed:
feat: add Amazon Linux 2023 support and update provider versions

BREAKING CHANGES:
- Terraform >= 1.5.7 (was >= 1.3.0)
- AWS Provider >= 5.0.0 (was >= 4.0.0)
- AWSCC Provider >= 1.0.0 (was >= 0.25.0)
- S3 Bucket Module ~> 5.0 (was >= 3.5.0)

Features:
- EC2 module: Use SSM parameters for AL2023-based Storage Gateway AMIs
- VMware module: Update OVA URL to v2 (AL2023-based)
- Add lifecycle rule to ignore AMI changes (prevents unexpected replacement)
- Enhanced EC2 metadata options with IMDSv2

prabirsekhri committed:
feat: security enhancements and module improvements

BREAKING CHANGE: disk_path is now required (no default)
- EC2: use disk_path = "/dev/nvme1n1"
- VMware: use disk_path = "/dev/sdb"

Security:
- Refactor security groups to use separate aws_security_group_rule resources
- Add egress rule for VPC endpoint security groups
- Convert for_each to count for single-resource conditionals

Improvements:
- Make domain_controllers optional for AD join
- Add non-sensitive gateway ARN and ID outputs
- Use null instead of empty string for optional variable defaults

prabirsekhri committed:
ci: add GitHub Actions workflow for static tests

- Add terraform validate, tflint, checkov, markdown-lint, terraform-docs
- Add configuration files for checkov, tflint, terraform-docs, mdl
- Remove tfsec (deprecated)

prabirsekhri committed:
docs: update examples and documentation

- Rename s3filegateway-vmware to s3-smb-filegateway-vmware
- Add Upgrading to v3.0.0 section in README
- Update all README files with terraform-docs
- Update CODEOWNERS with storage-gateway-contributors team
- Fix deprecated aws_region.current.name -> .id

prabirsekhri committed:
fix: use disk_node for VMware cache disk detection

- VMware examples now use disk_node='SCSI (0:1)' instead of disk_path
- disk_path remains for EC2 Nitro instances (/dev/nvme1n1)
- Both disk_path and disk_node are now optional with clear guidance

prabirsekhri committed:
docs: regenerate READMEs with terraform-docs

prabirsekhri committed:
style: terraform fmt

prabirsekhri commented:
https://github.com/prabirsekhri/terraform-aws-storagegateway/commit/48ca0ba02320fd0184a52b90abe023d760865b23 : functional tests run with Terraform cloud 

<img width="634" height="242" alt="image" src="https://github.com/user-attachments/assets/8ba7d663-690c-4130-87a4-007b2ff99006" />

SamuelOuvrard reviewed (approved):
LGTM 🚀

harsheetas reviewed (approved):
LGTM. Approve merging changes to main.

wellsiau-aws reviewed (approved):
LGTM
